File: frontend/src/lib/metrics.ts
import { formatEta, formatKwh } from "./format";
import { etaSecondsForRoute } from "./routes";
import type { ChargingSession, DockBay, EventLogItem, FleetMetric, Robot, Vehicle } from "./types";

export interface DerivedOperationsMetrics {
  commandBar: FleetMetric[];
  dockOccupancy: number;
  faultsToday: number;
  utilization: number;
}

export function deriveOperationsMetrics(
  robots: Robot[],
  vehicles: Vehicle[],
  sessions: ChargingSession[],
  events: EventLogItem[],
  energyToday: number,
  dockBays: DockBay[],
  missedCount = 0,
): DerivedOperationsMetrics {
  const available = robots.filter((robot) => (
    (robot.status === "idle" || robot.status === "docked")
    && !robot.assignedVehicleId
  )).length;
  const activeJobs = vehicles.filter((vehicle) => (
    vehicle.status === "assigned" || vehicle.status === "charging"
  )).length;
  const queueDepth = sessions.filter((session) => (
    session.status === "queued" || session.status === "interrupted"
  )).length;
  const enRouteEtas = robots
    .filter((robot) => robot.status === "en-route")
    .map((robot) => etaSecondsForRoute(robot.position, robot.route, robot.routeIndex));
  const averageEta = enRouteEtas.length > 0
    ? enRouteEtas.reduce((total, value) => total + value, 0) / enRouteEtas.length
    : null;
  const faultsToday = events.filter((event) => event.type === "fault").length;
  const healthyRobots = robots.filter((robot) => robot.status !== "faulted");
  const busyRobots = healthyRobots.filter((robot) => robot.status !== "idle" && robot.status !== "docked");
  const utilization = healthyRobots.length > 0 ? busyRobots.length / healthyRobots.length * 100 : 0;
  const dockOccupancy = Math.min(dockBays.length, robots.filter((robot) => robot.dockBayId).length);
  const carsInGarage = vehicles.filter((v) => v.status !== "departed").length;

  return {
    commandBar: [
      { id: "fleet", label: "Fleet online", value: `${healthyRobots.length}/${robots.length}` },
      { id: "available", label: "Available", value: `${available}` },
      { id: "active", label: "Jobs active", value: `${activeJobs}` },
      { id: "queue", label: "Queue depth", value: `${queueDepth}` },
      { id: "energy", label: "Energy today", value: formatKwh(energyToday) },
      { id: "eta", label: "Avg ETA", value: formatEta(averageEta) },
      { id: "cars", label: "Cars in garage", value: `${carsInGarage}` },
      { id: "missed", label: "Missed", value: missedCount > 0 ? `${missedCount}` : "—" },
    ],
    dockOccupancy,
    faultsToday,
    utilization,
  };
}

File: backend/app/metrics.py
"""Derived operational metrics."""

from __future__ import annotations

from app.routing import eta_seconds_for_route
from app.schemas import EventLogItem, Metrics, RobotStatus, SessionStatus, SystemState, VehicleStatus


def derive_metrics(state: SystemState) -> Metrics:
    robots = state.robots
    vehicles = state.vehicles
    sessions = state.sessions
    events = state.events

    healthy = [r for r in robots if r.status != RobotStatus.offline and r.status != RobotStatus.faulted]
    available = [r for r in robots if r.status in (RobotStatus.idle, RobotStatus.docked) and not r.assigned_vehicle_id]
    active_jobs = len([s for s in sessions if s.status in (SessionStatus.assigned, SessionStatus.en_route, SessionStatus.active)])
    queue_depth = len([s for s in sessions if s.status in (SessionStatus.queued, SessionStatus.interrupted)])
    cars_in_garage = len([v for v in vehicles if v.status != VehicleStatus.departed])
    faults_today = len([e for e in events if e.type == "fault"])
    missed = len([s for s in sessions if s.status == SessionStatus.missed])

    etas = [
        eta_seconds_for_route(r.position, r.route, r.route_index)
        for r in robots if r.status == RobotStatus.en_route
    ]
    avg_eta = sum(etas) / len(etas) if etas else None

    docked = len([
        r for r in robots
        if r.dock_bay_id and r.status in (RobotStatus.docked, RobotStatus.idle)
    ])
    dock_total = len(state.dock_bays)

    energy = sum(s.delivered_energy_kwh for s in sessions if s.status in (SessionStatus.active, SessionStatus.completed))

    return Metrics(
        fleet_online=len(healthy),
        robots_available=len(available),
        jobs_active=active_jobs,
        queue_depth=queue_depth,
        cars_in_garage=cars_in_garage,
        energy_today_kwh=round(energy, 1),
        average_eta_seconds=avg_eta,
        dock_occupancy=f"{docked}/{dock_total}",
        faults_today=faults_today,
        missed_requests=missed,
    )
